### PR TITLE
Simplify end/begin/rend/rbegin

### PR DESCRIPTION
**Pull request by dschwen**
Closes roystgnr/MetaPhysicL#79
Closes roystgnr/MetaPhysicL#78

### CODE REVIEW

File: src/utilities/include/metaphysicl/dynamic_std_array_wrapper.h [hunk 1]
```diff
@@ -128,13 +128,13 @@ class DynamicStdArrayWrapper
   reverse_iterator rbegin()
   {
     metaphysicl_assert(_dynamic_n <= N);
-    return _data.rbegin() + (N - _dynamic_n);
+    return _data.rend() + _dynamic_n;
```
Comment (roystgnr) [inline src/utilities/include/metaphysicl/dynamic_std_array_wrapper.h:131]: Can this be right?  Isn't rend() a reverse iterator?  You're not going to get to rbegin() from it by adding a positive number, are you?

I think we need some unit test coverage of these either way.

Comment (dschwen) [reply src/utilities/include/metaphysicl/dynamic_std_array_wrapper.h:131]: you use a reverse iterator like this:

```
for (std::vector<int>::reverse_iterator it = v.rbegin(); it != v.rend(); ++it)
```

i.e. the iterator is still advanced with `++`

Comment (dschwen) [reply src/utilities/include/metaphysicl/dynamic_std_array_wrapper.h]: Well, what can I say. _Of course_ you were right. It's `- _dynamic_n`! I added a unit test.

File: test/run_unit_tests.sh [hunk 2]
```diff
@@ -0,0 +1,31 @@
+#!/bin/sh
```
Comment (roystgnr) [inline test/run_unit_tests.sh:1]: This shouldn't have been committed; instead change run_unit_tests.sh.in and commit that?  This file should be written in our build dir by configure, not in our source dir manually.

**Review by roystgnr — CHANGES_REQUESTED** [review]

**Review by roystgnr — APPROVED** [review]

**Comment by lindsayad** [comment]
Well on a VTB test case that uses AD for everything, this did not make a notable difference. But I'm still glad we made the change. With this in, I'm thinking we should do a MetaPhysicL update in libMesh. @roystgnr thoughts?

**Comment by roystgnr** [comment]
I've got https://github.com/libMesh/TIMPI/pull/90 open for review while I work on some other things, and once I give that a little more time and testing I'll want to merge it all the way downstream, which means it'd be a great time to get MetaPhysicL downstream as well, yeah.  I wouldn't say this PR is worth an update by itself (I love that the code is cleaner, but I wouldn't be surprised if you could plug this into godbolt and see that compilers optimize to the same assembly before and after...), but the throw-on-excess-size is definitely worth pushing down.  Nothing else you want to get in before we do?

**Comment by lindsayad** [comment]
No, this was the last thing I wanted to see done. I know it's important for @dschwen to get the `protected` change. Also he and I will both appreciate the fixed assertion/throw checks, and as you say a lot of our users will appreciate the change to `throw`

**Pull request merged**